File: drivers/typescript/src/jwks.ts
import { createPublicKey, verify as verifySignature } from "node:crypto";

import { AuthError } from "./errors.js";

type JwtHeader = {
  alg?: string;
  kid?: string;
};

type JwtPayload = {
  exp?: number;
};

type Jwk = {
  kty: string;
  alg?: string;
  use?: string;
  kid: string;
  n: string;
  e: string;
};

type JwksDocument = {
  keys: Jwk[];
};

function decodeBase64Url(input: string): Buffer {
  const normalized = input.replace(/-/g, "+").replace(/_/g, "/");
  const pad = normalized.length % 4;
  const padded = pad === 0 ? normalized : normalized + "=".repeat(4 - pad);
  return Buffer.from(padded, "base64");
}

function parseJsonSegment<T>(segment: string): T {
  return JSON.parse(decodeBase64Url(segment).toString("utf8")) as T;
}

export class JwksCache {
  private url: string;
  private byKid = new Map<string, Jwk>();

  constructor(url: string) {
    this.url = url;
  }

  async verifyAccessToken(token: string): Promise<void> {
    const parts = token.split(".");
    if (parts.length !== 3) {
      throw new AuthError("invalid JWT format");
    }

    const [headerSeg, payloadSeg, sigSeg] = parts;
    const header = parseJsonSegment<JwtHeader>(headerSeg);
    const payload = parseJsonSegment<JwtPayload>(payloadSeg);

    if (header.alg !== "RS256") {
      throw new AuthError(`unsupported JWT alg '${header.alg ?? "unknown"}'`);
    }
    if (!header.kid) {
      throw new AuthError("JWT header missing kid");
    }

    if (!this.byKid.has(header.kid)) {
      await this.refresh();
    }
    if (!this.byKid.has(header.kid)) {
      await this.refresh();
    }

    const jwk = this.byKid.get(header.kid);
    if (!jwk) {
      throw new AuthError(`kid '${header.kid}' not found in JWKS`);
    }

    const signingInput = Buffer.from(`${headerSeg}.${payloadSeg}`, "utf8");
    const signature = decodeBase64Url(sigSeg);
    const publicKey = createPublicKey({ key: jwk, format: "jwk" });

    const valid = verifySignature("RSA-SHA256", signingInput, publicKey, signature);
    if (!valid) {
      throw new AuthError("JWT RS256 signature verification failed");
    }

    if (typeof payload.exp === "number") {
      const now = Math.floor(Date.now() / 1000);
      if (payload.exp <= now) {
        throw new AuthError("JWT token expired");
      }
    }
  }

  private async refresh(): Promise<void> {
    const res = await fetch(this.url);
    if (!res.ok) {
      throw new AuthError(`JWKS fetch failed with HTTP ${res.status}`);
    }

    const doc = (await res.json()) as JwksDocument;
    this.byKid.clear();

    for (const key of doc.keys ?? []) {
      if (key.kty !== "RSA") {
        continue;
      }
      if (key.alg && key.alg !== "RS256") {
        continue;
      }
      if (key.use && key.use !== "sig") {
        continue;
      }
      this.byKid.set(key.kid, key);
    }
  }
}

File: grumpydb-client/src/jwks.rs
use std::collections::HashMap;

use jsonwebtoken::{Algorithm, DecodingKey, Validation, decode, decode_header};
use serde::Deserialize;
use serde_json::Value;

use crate::ClientError;

#[derive(Debug, Clone, Deserialize)]
struct JwksDocument {
    keys: Vec<Jwk>,
}

#[derive(Debug, Clone, Deserialize)]
struct Jwk {
    kty: String,
    alg: Option<String>,
    #[serde(rename = "use")]
    use_: Option<String>,
    kid: String,
    n: String,
    e: String,
}

#[derive(Default)]
pub(crate) struct JwksCache {
    url: Option<String>,
    by_kid: HashMap<String, DecodingKey>,
}

impl JwksCache {
    pub(crate) fn set_url(&mut self, url: impl Into<String>) {
        self.url = Some(url.into());
        self.by_kid.clear();
    }

    pub(crate) fn configured(&self) -> bool {
        self.url.is_some()
    }

    pub(crate) async fn verify_access_token(&mut self, token: &str) -> Result<(), ClientError> {
        let header = decode_header(token)
            .map_err(|e| ClientError::Jwt(format!("invalid token header: {e}")))?;

        if header.alg != Algorithm::RS256 {
            return Err(ClientError::Jwt(format!(
                "unsupported JWT alg {:?}, expected RS256",
                header.alg
            )));
        }

        let kid = header
            .kid
            .ok_or_else(|| ClientError::Jwt("token header missing kid".into()))?;

        if !self.by_kid.contains_key(&kid) {
            self.refresh().await?;
        }

        if !self.by_kid.contains_key(&kid) {
            self.refresh().await?;
        }

        let key = self
            .by_kid
            .get(&kid)
            .ok_or_else(|| ClientError::Jwt(format!("kid '{kid}' not found in JWKS")))?;

        let mut validation = Validation::new(Algorithm::RS256);
        validation.required_spec_claims.insert("exp".to_string());

        decode::<Value>(token, key, &validation)
            .map_err(|e| ClientError::Jwt(format!("token verification failed: {e}")))?;
        Ok(())
    }

    async fn refresh(&mut self) -> Result<(), ClientError> {
        let url = self
            .url
            .clone()
            .ok_or_else(|| ClientError::Jwt("JWKS URL not configured".into()))?;

        let doc: JwksDocument = reqwest::get(&url)
            .await
            .map_err(|e| ClientError::Jwt(format!("JWKS fetch failed: {e}")))?
            .error_for_status()
            .map_err(|e| ClientError::Jwt(format!("JWKS HTTP error: {e}")))?
            .json()
            .await
            .map_err(|e| ClientError::Jwt(format!("invalid JWKS payload: {e}")))?;

        self.by_kid.clear();
        for key in doc.keys {
            if key.kty != "RSA" {
                continue;
            }
            if let Some(alg) = &key.alg
                && alg != "RS256"
            {
                continue;
            }
            if let Some(use_) = &key.use_
                && use_ != "sig"
            {
                continue;
            }

            let decoding = DecodingKey::from_rsa_components(&key.n, &key.e)
                .map_err(|e| ClientError::Jwt(format!("invalid RSA key in JWKS: {e}")))?;
            self.by_kid.insert(key.kid, decoding);
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_jwks_cache_not_configured_by_default() {
        let cache = JwksCache::default();
        assert!(!cache.configured());
    }

    #[test]
    fn test_jwks_cache_configured_after_url_set() {
        let mut cache = JwksCache::default();
        cache.set_url("http://127.0.0.1:8080/.well-known/jwks.json");
        assert!(cache.configured());
    }
}
